### PR TITLE
Add Concat operator with CPU kernel 

### LINKED ISSUE #3772: Concat Operator
TODO
- [x] CPU Kernel, #3875 
- [ ] Use Split Op to compute gradient
- [ ] GPU Kernel

### LINKED ISSUE #3875: Concat Op with CPU kernel

### PR DESCRIPTION
**Pull request by Yancey1989**
Fixed #3875

### CODE REVIEW

File: paddle/operators/concat_op.cc [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+class ConcatOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+ protected:
+  void InferShape(const framework::InferShapeContext &ctx) const override {
+    auto ins = ctx.MultiInput<framework::Tensor>("X");
+    auto *out = ctx.Output<framework::Tensor>("Out");
+    const int axis = static_cast<int>(ctx.op_.GetAttr<int>("axis"));
```
Comment (qingqing01) [inline paddle/operators/concat_op.cc]: The `InferShapeContext` encapsulated the function about how to get input and output variable, but except attributes. Whether the attr's get-functions should be included in `InferShapeContext`?  @reyoung @jacquesqiao

File: paddle/operators/concat_op.cc [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    int N = ins.size();
+
+    PADDLE_ENFORCE_GT(N, 1, "Input tensors count should >= 1.");
```
Comment (qingqing01) [inline paddle/operators/concat_op.cc]: The count of input variables should > 1. In only one input, there is no need to concatenate.

File: paddle/operators/concat_op.cc [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+    auto dim_zero = ins[0]->dims();
+    auto dim_zero_size = dim_zero.size();
+    auto concat_dim = dim_zero;
+    for (int i = 1; i < N; i++) {
+      PADDLE_ENFORCE_EQ(dim_zero_size, ins[i]->dims().size(),
+                        "input tensors should have the same dim size.");
```
Comment (qingqing01) [inline paddle/operators/concat_op.cc]: the same rank
dim size is a little confused

File: python/paddle/v2/framework/tests/test_concat_op.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+    def setUp(self):
+        self.type = "concat"
+        self.inputs = {
+            'X': [
+                np.random.random((16, 32)).astype("float32"), np.random.random(
+                    (8, 32)).astype("float32"), np.random.random(
+                        (2, 32)).astype("float32")
+            ]
+        }
```
Comment (qingqing01) [inline python/paddle/v2/framework/tests/test_concat_op.py]: Maybe short lines as follows:
``` 
x1 =  np.random.random((16, 32)).astype("float32")
x2 = np.random.random((8, 32)).astype("float32")
x3 = np.random.random (2, 32)).astype("float32")
self.inputs = {'X': [x1, x2, x3]}
```

File: paddle/operators/CMakeLists.txt [hunk 5]
```diff
@@ -73,3 +73,4 @@ op_library(uniform_random_op SRCS uniform_random_op.cc uniform_random_op.cu)
 op_library(lookup_table_op SRCS lookup_table_op.cc lookup_table_op.cu)
 op_library(scale_op SRCS scale_op.cc scale_op.cu DEPS net_op)
 op_library(minus_op SRCS minus_op.cc minus_op.cu DEPS scale_op)
+op_library(concat_op SRCS concat_op.cc concat_op.cu DEPS math_function)
```
[the remark this answers is not in the document]
Comment (Yancey1989) [reply paddle/operators/CMakeLists.txt]: Done.

File: paddle/operators/concat_op.cc (hunk 2 above)
Comment (Yancey1989) [reply paddle/operators/concat_op.cc]: Done, sorry my comment error.

File: paddle/operators/concat_op.cc (hunk 3 above)
Comment (Yancey1989) [reply paddle/operators/concat_op.cc]: Done, we only need to keep the same dim except the axis.

File: python/paddle/v2/framework/tests/test_concat_op.py (hunk 4 above)
Comment (Yancey1989) [reply python/paddle/v2/framework/tests/test_concat_op.py]: Done. 
But the inputs can not be a Python list or `np.array`, and I create a issue #3805 to describe this problem.

File: paddle/operators/concat_op.cc (hunk 1 above)
Comment (jacquesqiao) [reply paddle/operators/concat_op.cc]: I think get attr from op is fine, but should add a GetOp to ctx, I will add it.

File: paddle/operators/concat_op.cc [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+class ConcatOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+ protected:
+  void InferShape(const framework::InferShapeContext &ctx) const override {
+    auto ins = ctx.MultiInput<framework::Tensor>("X");
+    auto *out = ctx.Output<framework::Tensor>("Out");
+    auto axis = static_cast<int>(ctx.GetAttr<int>("axis"));
+    int N = ins.size();
```
Comment (typhoonzero) [inline paddle/operators/concat_op.cc]: You can use more `size_t`

File: paddle/operators/concat_op.cc [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+};
+
+class ConcatOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  ConcatOpMaker(framework::OpProto *proto, framework::OpAttrChecker *op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput("X", "the input tensors of concat operator.").AsDuplicable();
+    AddOutput("Out", "the output tensor of concat operator.");
+    AddComment(R"DOC(
+            Join the input tensors alone the with the axis.
```
Comment (typhoonzero) [inline paddle/operators/concat_op.cc]: Join the input tensors along with the axis.

File: paddle/operators/concat_op.cc [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+        )DOC");
```
Comment (typhoonzero) [inline paddle/operators/concat_op.cc]: May be add some samples like concat matrices

File: paddle/operators/concat_op.cc [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    AddAttr<int>("axis", "The axis alone which the inputs will be joined")
```
Comment (typhoonzero) [inline paddle/operators/concat_op.cc]: Remove "alone"

File: paddle/operators/concat_op.cu [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#define EIGEN_USE_GPU
+#include "paddle/operators/concat_op.h"
+
+namespace ops = paddle::operators;
+REGISTER_OP_GPU_KERNEL(concat,
```
Comment (typhoonzero) [inline paddle/operators/concat_op.cu]: The ConcatKernel seems only for CPU, do not register for GPU.

File: paddle/operators/concat_op.h [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    }
+    int output_offset = 0;
+    out->mutable_data<T>(platform::CPUPlace());
+    for (int i = 0; i < N; i++) {
+      auto& in = ins[i];
+      auto axis_dim = in->dims()[axis];
+      for (int j = 0; j < before; j++) {
+        int len = axis_dim * after * sizeof(T);
+        const T* src = in->data<T>() + axis_dim * after * j;
+        T* dest = out->data<T>() + output_offset + output_axis_dim * after * j;
```
Comment (typhoonzero) [inline paddle/operators/concat_op.h]:
```c++
T* out_data = out->mutable_data<T>(platform::CPUPlace());
T* dest = out_data + ...
```

File: paddle/operators/concat_op.cc [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+ protected:
+  void InferShape(const framework::InferShapeContext &ctx) const override {
+    auto ins = ctx.MultiInput<framework::Tensor>("X");
+    auto *out = ctx.Output<framework::Tensor>("Out");
+    auto axis = static_cast<int>(ctx.GetAttr<int>("axis"));
+    int N = ins.size();
+
+    PADDLE_ENFORCE_GT(N, 1, "Input tensors count should > 1.");
+
+    auto dim_zero = ins[0]->dims();
```
Comment (typhoonzero) [inline paddle/operators/concat_op.cc]: `dim_zero` is odd, rename to something like `input_dim`?

Comment (Yancey1989) [reply paddle/operators/concat_op.cc]: Done.

File: paddle/operators/concat_op.cc [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+
+class ConcatOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+ protected:
+  void InferShape(const framework::InferShapeContext &ctx) const override {
+    auto ins = ctx.MultiInput<framework::Tensor>("X");
+    auto *out = ctx.Output<framework::Tensor>("Out");
+    auto axis = static_cast<int>(ctx.GetAttr<int>("axis"));
```
[the remark this answers is not in the document]
Comment (Yancey1989) [reply paddle/operators/concat_op.cc]: Done.

File: paddle/operators/concat_op.cc (hunk 6 above)
Comment (Yancey1989) [reply paddle/operators/concat_op.cc]: Done.

File: paddle/operators/concat_op.cu (hunk 10 above)
Comment (Yancey1989) [reply paddle/operators/concat_op.cu]: Done. I'll add GPU kernel in next PR.

File: paddle/operators/concat_op.cc (hunk 9 above)
Comment (Yancey1989) [reply paddle/operators/concat_op.cc]: Done.

File: paddle/operators/concat_op.cc (hunk 8 above)
Comment (Yancey1989) [reply paddle/operators/concat_op.cc]: Done.

File: paddle/operators/concat_op.cc (hunk 7 above)
Comment (Yancey1989) [reply paddle/operators/concat_op.cc]: Done.

File: paddle/operators/concat_op.cc [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+class ConcatOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+ protected:
+  void InferShape(const framework::InferShapeContext &ctx) const override {
+    auto ins = ctx.MultiInput<framework::Tensor>("X");
+    auto *out = ctx.Output<framework::Tensor>("Out");
+    int axis = static_cast<int>(ctx.Attr<int>("axis"));
+    size_t N = ins.size();
```
Comment (typhoonzero) [inline paddle/operators/concat_op.cc]: https://google.github.io/styleguide/cppguide.html#General_Naming_Rules

No capitalized vars.

File: paddle/operators/concat_op.cc [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+
+    PADDLE_ENFORCE_GT(N, 1, "Input tensors count should > 1.");
+
+    auto out_dims = ins[0]->dims();
+    int in_zero_dims_size = out_dims.size();
```
Comment (typhoonzero) [inline paddle/operators/concat_op.cc]: why not uniform to size_t?

Comment (Yancey1989) [reply paddle/operators/concat_op.cc]: Done.

File: paddle/operators/concat_op.cc (hunk 14 above)
Comment (Yancey1989) [reply paddle/operators/concat_op.cc]: Done.

**Pull request merged**